sound/soc/intel/boards/sof_sdw_rt_amp.c

```diff
@@ -183,16 +183,34 @@ int rt_amp_spk_rtd_init(struct snd_soc_pcm_runtime *rtd)
 	const struct snd_soc_dapm_route *rt_amp_map;
 	char codec_name[CODEC_NAME_SIZE];
 	struct snd_soc_dai *dai;
+	const char *c_temp;
 	int ret;
 	int i;
 
 	rt_amp_map = get_codec_name_and_route(rtd, codec_name);
 
-	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
-					  "%s spk:%s",
-					  card->components, codec_name);
-	if (!card->components)
-		return -ENOMEM;
+	/* Add components string for each codec dai */
+	for_each_rtd_codec_dais(rtd, i, dai) {
+		/*
+		 * rt_amp_spk_rtd_init() only for rt13xx amps. rt7xx multi-function codecs will
+		 * use their own rtd_init function.
```

Comment:
this comment confused me a lot...
multi-function codecs, like rt7xx will not use the `rt_amp_spk_rtd_init` but their own rtd_init function since they need to add slightly different routes (I guess).
So this comment and the check for name might be redundant (and confusing)?

Reply:
@ujfalusi "spk:rt7xx" will be added by `rt7xx_spk_rtd_init`, and we don't want to add it here. And here we add "spk:%s" for each codec dai, so we need to check the dai->name and only add "spk:rt13xx" to card->components.

Reply:
But `spk:rt7xx` is added by `rt7xx_spk_rtd_init` (where xx is 12 or 22 atm, right?) and in those case the `rt_amp_spk_rtd_init` will not be called, or am I missing something?
For me it looks like that `rt_amp_spk_rtd_init()` will never be called with `rt7-something-`

Reply:
We could have rt7xx + rt13xx combination case. For example, rt712 + rt1316 in a single dai link. In that case, rt1316 will call `rt_amp_spk_rtd_init` and `rt_amp_spk_rtd_init` will check all codec dais in the dai link which are rt712 and rt1316.
IOW, `rt_amp_spk_rtd_init` is called by rt13xx, but it will check all amps in the dai link which may contain rt7xx. And we need to skip the non-rt13xx amps in the dai link.

Reply:
I don't get this (I got what you are saying, but)...
if we have multiple rt13xx in a single dai link then we add the entries multiple times?
The `rt_amp_spk_rtd_init` will be called for each dais, no?
If I look closer, the other `spk_rtd_init`s are adding a single entry unconditionally, they don't scan for all dais.

I'm sure this works, but it is a bit holistic for me ;)

Reply:
> I don't get this (I got what you are saying, but)... if we have multiple rt13xx in a single dai link then we add the entries multiple times? The `rt_amp_spk_rtd_init` will be called for each dais, no? If I look closer, the other `spk_rtd_init`s are adding a single entry unconditionally, they don't scan for all dais.
> 

rt13xx will not be added multiple times, we check it with
		if (strstr(card->components, c_temp))
			continue;

rt7xx add a single entry unconditionally because the rt7xx_spk_rtd_init is specific for the codec, and we don't expect that there will be more than one rt7xx in a link.
However, rt_amp_spk_rtd_init is for all rt13xx amps, and it is possible we combine two different rt13xx amps like rt1316 + rt1308 together. It does not likely happen, but it is harmless to support this corner case.

And I am thinking maybe someday we can use a single spk_rtd_init for all kinds of amp. And we definitely need to check all codecs in the dai lnik.

> I'm sure this works, but it is a bit holistic for me ;)

Reply:
@bardliao, right, I see. What is confusing is the fact that the `sof_sdw_dai_info.rtd_init()` callback has wrong name and scope (imho), it should be
```
int (*dai_init)(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai);
bool dai_init_done; /* Indicate that the dai_init callback is done */
```
and deal with the given dai only.

Reply:
> @bardliao, right, I see. What is confusing is the fact that the `sof_sdw_dai_info.rtd_init()` callback has wrong name and scope (imho), it should be
> 
> ```
> int (*dai_init)(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai);
> bool dai_init_done; /* Indicate that the dai_init callback is done */
> ```
> 
> and deal with the given dai only.

Good suggestion, I will do it later.